apps/playwright-browser-tunnel/src/PlaywrightBrowserTunnel.ts

```diff
@@ -0,0 +1,524 @@
+        '--all'
+      ]);
+    } catch (error) {
+      this._terminal.writeLine(
+        `Failed to uninstall browsers: ${error instanceof Error ? error.message : error}`
+      );
+    }
+
+    await this.cleanTempFilesAsync();
+  }
+
+  private async _runCommandAsync(command: string, args: string[]): Promise<void> {
+    const tmpPath: string = this._tmpPath;
+    await FileSystem.ensureFolderAsync(tmpPath);
+    this._terminal.writeLine(`Running command: ${command} ${args.join(' ')} in ${tmpPath}`);
+
+    const cp: ChildProcess = Executable.spawn(command, args, {
+      stdio: [
+        'ignore', // stdin
+        'pipe', // stdout
+        'pipe' // stderr
+      ],
+      currentWorkingDirectory: tmpPath
+    });
+
+    cp.stdout?.pipe(
+      new TerminalStreamWritable({
+        terminal: this._terminal,
+        severity: TerminalProviderSeverity.log
+      })
+    );
+    cp.stderr?.pipe(
+      new TerminalStreamWritable({
+        terminal: this._terminal,
+        severity: TerminalProviderSeverity.error
+      })
+    );
+
+    await Executable.waitForExitAsync(cp);
+  }
+
+  private async _installPlaywrightCoreAsync({
+    playwrightVersion
+  }: Pick<IHandshake, 'playwrightVersion'>): Promise<void> {
+    this._terminal.writeLine(`Installing playwright-core version ${playwrightVersion}`);
+    await this._runCommandAsync('npm', [
+      'install',
+      `playwright-core-${playwrightVersion}@npm:playwright-core@${playwrightVersion}`
+    ]);
+  }
+
+  private async _installPlaywrightBrowsersAsync({
+    playwrightVersion,
+    browserName
+  }: Pick<IHandshake, 'playwrightVersion' | 'browserName'>): Promise<void> {
+    await this._installPlaywrightCoreAsync({ playwrightVersion });
+    this._terminal.writeLine(`Executing playwright-core version ${playwrightVersion}`);
+    await this._runCommandAsync('node', [
+      `node_modules/playwright-core-${playwrightVersion}/cli.js`,
+      'install',
+      browserName
+    ]);
+  }
+
+  private _tryConnectAsync(): Promise<WebSocket> {
+    const wsEndpoint: string | undefined = this._wsEndpoint;
+    if (!wsEndpoint) {
+      return Promise.reject(new Error('WebSocket endpoint is not defined'));
+    }
+    return new Promise<WebSocket>((resolve, reject) => {
+      const ws: WebSocket = new WebSocket(wsEndpoint);
+      ws.on('open', () => {
+        this._terminal.writeLine(`WebSocket connection opened`);
+        resolve(ws);
+      });
+      ws.on('error', (error) => {
+        reject(error);
+      });
+    });
+  }
+
+  // TODO: Only supporting one test at a time.
+  // Need to support multiple simultaneous connections for parallel tests.
+  private async _pollConnectionAsync(): Promise<WebSocket> {
+    this._terminal.writeLine(`Waiting for WebSocket connection`);
+    return new Promise((resolve, reject) => {
+      const interval: NodeJS.Timeout = setInterval(async () => {
+        try {
+          const ws: WebSocket = await this._tryConnectAsync();
```

Comment:
If `_tryConnectAsync` takes longer than 500ms to execute and has at least one async tick internally, then it can get called twice concurrently. Is that handled?